File: baekjoon/graph/1717/union_find.cpp
#include <iostream>

const int   MAX = 1000001;

int parents[MAX];

int getParent(int x)
{
    if (parents[x] == x)
        return x;
    return parents[x] = getParent(parents[x]);
}

void union_(int x, int y)
{
    int xParent = getParent(x);
    int yParent = getParent(y);

    if (xParent > yParent)
        parents[xParent] = yParent;
    else
        parents[yParent] = xParent;
}

int main(void)
{
    std::cin.tie(0)->sync_with_stdio(0);

    int cmd, x, y;
    int vCnt, cmdCnt;

    std::cin >> vCnt >> cmdCnt;
    ++vCnt;
    for (int v = 1; v < vCnt; ++v)
        parents[v] = v;

    while (cmdCnt--)
    {
        std::cin >> cmd >> x >> y;
        if (cmd == 0)
            union_(x, y);
        else
            std::cout << (getParent(x) == getParent(y) ? "YES" : "NO") << '\n';
    }
    return (0);
}

File: baekjoon/graph/1717/union_find.ts
const lines = require('fs').readFileSync(0).toString().trim().split('\n');
const [vCnt, cmdCnt] = lines[0].split(' ').map(Number);

function print_ans()
{
    const parents : number[] = Array(vCnt + 1).fill(0);
    
    for (var i = 1; i <= vCnt; ++i)
        parents[i] = i;

    function getParent(x: number)
    {
        if (parents[x] == x)
            return x;
        return parents[x] = getParent(parents[x]);
    }

    function union(x: number, y: number)
    {
        const xParent = getParent(x);
        const yParent = getParent(y);

        if (xParent > yParent)
            parents[xParent] = yParent;
        else
            parents[yParent] = xParent;
    }

    var cmd, x, y;
    for (var i = 1; i < lines.length; ++i)
    {
        [cmd, x, y] = lines[i].split(' ').map(Number);
        if (cmd == 0)
            union(x, y);
        else
            console.log(getParent(x) == getParent(y) ? "YES" : "NO");
    }
}

print_ans();
